strawberry/types/type_resolver.py

```diff
@@ -222,36 +226,128 @@ def _resolve_types(fields: List[FieldDefinition]) -> List[FieldDefinition]:
+
+    Type #2a:
+        A field defined using strawberry.field as a function, but without
+        supplying a resolver. Again, a FieldDefinition will need to be created
+        in this function. Type annotation is required.
+
+    Type #2b:
+        A field defined using strawberry.field as a function, with a supplied
+        resolver.
+
+        The type hint is optional, but if supplied, it must match the return
+        type of the resolver. Type annnotation is required if resolver is not
+        type annotated; if both are annotated, they must match.
+
+        Implementation note: If both type annotations are provided, there will
+        be a redundant Type #1-style entry in the dataclass' field list. If the
+        strawberry.field call does not specify a `name`, the name of the field
+        on the class will be used.
+
+    Type #2c:
+        A field defined using @strawberry.field as a decorator around the
+        resolver. The resolver must be type-annotated.
+
+    Final `name` attribute priority:
+    1. Type #2 `name` attribute. This will be defined with an explicit
+       strawberry.field(name=...). No camelcase-ification will be done, as the
+       user has explicitly stated the field name.
+    2. Field name on the cls. Will exist for all fields other than Type #2c.
+       Field names will be converted to camelcase.
+    """
+    field_definitions: Dict[str, FieldDefinition] = {}
+
+    # Type #1 fields
+    type_1_fields: Dict[str, dataclasses.Field] = {
+        field.name: field for field in dataclasses.fields(cls)
+    }
+
+    # Type #2 fields
+    type_2_fields = {}
+    for field_name, field in cls.__dict__.items():
+        if hasattr(field, "_field_definition"):
+            type_2_fields[field_name] = field
+
+    for field_name, field in type_2_fields.items():
+
+        field_definition: FieldDefinition = field._field_definition
+
+        # Check if there is a matching Type #1 field:
+        if field_name in type_1_fields:
+            # Make sure field and resolver types are the same if both are
+            # defined
+            # TODO: https://github.com/strawberry-graphql/strawberry/issues/396
+            # assert field.type == resolver.type
+
+            # Grab the type from the field if the resolver has no type
+            if field_definition.type is None:
+                field_type = type_1_fields[field_name].type
+                field_definition.type = field_type
+
+            # Stop tracking the Type #1 field, an explict strawberry.field was
+            # defined
+            type_1_fields.pop(field_name)
+
+        # Otherwise, ensure that a resolver has been specified
+        else:
+            if field_definition.base_resolver is None:
+                # This should be caught by _wrap_dataclass in type.py, but just
+                # in case, we'll check again
+                raise MissingFieldAnnotationError(field_name)
```

Comment:
when would this happen? if it doesn't I'd skip this check 😊 
we can add a `assert field_definition.base_resolver` instead, so it is clear there was a bug :)

Reply:
If something else in the future ever calls `_get_fields` I figured it would be helpful to catch that issue here too

I'm a little torn here... The [entire `else` clause](https://github.com/strawberry-graphql/strawberry/blob/246d23d6998a33342186e9f7caa6ef96c422ec9d/strawberry/types/type_resolver.py#L315-L324) this is in currently does nothing but identify problems with the FieldDefinition. The [other item in here](https://github.com/strawberry-graphql/strawberry/blob/246d23d6998a33342186e9f7caa6ef96c422ec9d/strawberry/types/type_resolver.py#L321-L324) is also a check, but this time it _can_ currently get hit. It feels weird to have one be an `assert` and the other raise an exception, does it not?

Reply:
maybe, I just think they are different cases so it might be ok to have different ways to fail. One happens as a user error while using resolver, the other one, not sure :)

Reply:
up to you 👍

Reply:
let's keep it for now, I think if we do this: https://github.com/strawberry-graphql/strawberry/pull/429#discussion_r481392751

we might be good :)